### PR TITLE
[graph_trainer] Add precompile artifact serialization and loading

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #2646
* #2650
* #2644
* __->__ #2643
* #2642

Add precompile_save() and precompile_load() functions that serialize and
deserialize compiled AOT graphs using BundledAOTAutogradSerializableCallable.
Artifacts include the serialized compiled function, parameter/buffer specs,
input/output tree specs, and metadata. Deserialization is deferred to first
call so Triton kernels load on the correct CUDA device.

Includes config fingerprint validation: `compute_config_fingerprint()` hashes
model param/buffer names+shapes+dtypes, parallelism dimensions (world_size,
dp_replicate, dp_shard, cp, tp, pp, ep, etp), and compile config (mode,
backend, passes, joint_passes) into a 16-char SHA-256 digest. The fingerprint
is stored in the artifact on save and validated on load — mismatches produce
a clear error, and legacy artifacts without fingerprints log a warning.

@claude

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
## Loss Curve Validation

Tested on 8x H100 with `--debug.deterministic --debug.seed=42`, 50 steps.

The `_unwrap_serializable` helper correctly extracts the `BundledAOTAutogradSerializableCallable` from PyTorch's `aot_compile_joint_with_descriptors` function wrapper. This is needed because PyTorch wraps the serializable callable with `functools.wraps`, so the `__wrapped__` attribute points to the actual serializable object.

**Precompile save/load roundtrip produces bit-identical losses** (max |diff| = 0.0).

Full comparison

File: torchtitan/experiments/graph_trainer/tests/test_precompile.py [hunk 1]
```diff
@@ -60,5 +63,333 @@ def test_delete_nonexistent_noop(self):
...
+        params = [
+            ("layer.weight", torch.zeros(4, 4)),
+            ("layer.bias", torch.zeros(4)),
+        ]
+    if buffers is None:
+        buffers = [("running_mean", torch.zeros(4))]
+
+    model = MagicMock()
+    model.named_parameters.side_effect = lambda: iter(params)
+    model.named_buffers.side_effect = lambda: iter(buffers)
```
Comment (bobrenjc93) [inline torchtitan/experiments/graph_trainer/tests/test_precompile.py:126]: can you add a comment saying why you do this?

File: torchtitan/experiments/graph_trainer/precompile.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        f"Precompile artifact loaded: key={artifact_key}, "
+        f"params={len(artifact.params_spec)}, "
+        f"buffers={len(artifact.buffers_spec)}, "
+        f"fingerprint={artifact.config_fingerprint}, "
+        f"metadata={artifact.metadata}"
+    )
+
+    out_spec = artifact.out_spec
+    serialized_fn_bytes = artifact.serialized_fn
+    compiled_fn_holder: list[Callable] = []
```
Comment (bobrenjc93) [inline torchtitan/experiments/graph_trainer/precompile.py]: Why is this a list instead of an Optional?
